### PR TITLE
PD-627 proxy volumes

### PR DESCRIPTION
**Pull request by nathan-px**

### CODE REVIEW

File: content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
+---
+title: Create proxy volume PVCs
+weight: 4
+keywords: create proxy volumes, PVC, kubernetes, k8s
+description: 
+series: k8s-vol
+---
+
+Portworx proxy volumes reflect an external data source onto a Portworx volume. The actual data for these volumes resides on the external data source and does not consume any storage from the Portworx storage pools.
```
Comment (adityadani) [inline content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md:9]: nit: lets use the term proxy instead of reflect. So `Portworx proxy volumes proxy an external data source...`

File: content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+You can use proxy volumes to reflect an external NFS share onto your volumes. If you have an NFS share residing outside of a Kubernetes cluster and you wish to access it within an application pod, you can create a Portworx proxy volume that points to this external NFS share. Portworx acts as a medium and makes the external NFS data available to the pods running in Kubernetes.
+
+Portworx uses the host’s NFS utilities to mount the external NFS share when a Pod using the proxy-volume PVC gets scheduled on a node. Depending on how you configure it, you can mount an entire NFS share to volumes, or you can mount only portions of the NFS share to volumes by specifying a directory to specific sub-path.
+
+## Access an external NFS share
+
+You can attach the full NFS share to Portworx as a proxy volume. Applications will place volumes at the root path of the NFS share endpoint provided. <!-- not sure if this is accurate, need to verify. When would someone use this, and when would they use the sub-path? --> The examples in these instructions create a proxy volume for an nginx container.
```
Comment (nathan-px) [inline content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md:17]: @adityadani I have a number of questions as inline comments, find them by searching `<!--`

File: content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+title: Create proxy volume PVCs
+weight: 4
+keywords: create proxy volumes, PVC, kubernetes, k8s
+description: 
+series: k8s-vol
+---
+
+Portworx proxy volumes reflect an external data source onto a Portworx volume. The actual data for these volumes resides on the external data source and does not consume any storage from the Portworx storage pools.
+
+You can use proxy volumes to reflect an external NFS share onto your volumes. If you have an NFS share residing outside of a Kubernetes cluster and you wish to access it within an application pod, you can create a Portworx proxy volume that points to this external NFS share. Portworx acts as a medium and makes the external NFS data available to the pods running in Kubernetes.
```
Comment (adityadani) [inline content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md:11]: same here

File: content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+Portworx uses the host’s NFS utilities to mount the external NFS share when a Pod using the proxy-volume PVC gets scheduled on a node. Depending on how you configure it, you can mount an entire NFS share to volumes, or you can mount only portions of the NFS share to volumes by specifying a directory to specific sub-path.
```
Comment (adityadani) [inline content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md:13]: nit: by specifying a directory sub-path.

File: content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md (hunk 2 above)
Comment (adityadani) [inline content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md:17]: Attach is usually tied to block volumes. So lets rephrase this to
You can access a full NFS share in Portworx as a proxy volume.

Comment (adityadani) [inline content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md:17]: You can say this: 

Application using this spec will have access to the whole NFS share. If you wish to access only a sub directory within an NFS share go to next section.

File: content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+
+## Access an external NFS share
+
+You can attach the full NFS share to Portworx as a proxy volume. Applications will place volumes at the root path of the NFS share endpoint provided. <!-- not sure if this is accurate, need to verify. When would someone use this, and when would they use the sub-path? --> The examples in these instructions create a proxy volume for an nginx container.
+
+1. Create a storage class spec for proxy volumes, specifying your own values for the following:
+
+  * **parameters.proxy-volume_endpoint:** With the endpoint of the external NFS share Portworx is reflecting from.
```
Comment (adityadani) [inline content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md:21]: nit: the params are these:
```
    proxy_endpoint: "nfs:<nfs-share-endpoint>"
    proxy_nfs_exportpath: "/<mount-path>"
```

File: content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+5. Apply the spec:
+
+  ```text 
+  kubectl create -f <pod-name>.yaml
+  ```
+
+## Accessing a sub-path of an external NFS share
+
+You can attach a sub-path of an NFS share to Portworx as a proxy volume. Under this approach, applications will place volumes at the specified sub-path of the NFS share endpoint provided, allowing you to use only part of your NFS share. <!-- not sure if this is accurate, need to verify. --> The examples in these instructions create a proxy volume for an nginx container.
```
Comment (adityadani) [inline content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md:111]: same here: lets not use attach.
also you could say:
Under this approach applications will have access to a specific sub path within the NFS share.

File: content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+  * **parameters.proxy-volume_endpoint:** With the export path on the NFS server. 
+
+  ```text
+  kind: StorageClass
+  apiVersion: storage.k8s.io/v1
+  metadata:
+    name: portworx-proxy-volume-volume
+  provisioner: kubernetes.io/portworx-volume
+  parameters:
+    proxy-volume_endpoint: "nfs:<nfs-share-endpoint>"
```
Comment (adityadani) [inline content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md:128]: same here:
```
    proxy_endpoint: "nfs:<nfs-share-endpoint>"
    proxy_nfs_exportpath: "/<mount-path>"
```

File: content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+
+2. Apply the spec:
+
+  ```text
+  kubectl create -f <storageclass-name>.yaml
+  ```
+
+3. Create the Portworx proxy volume PVC spec, specifying your own values for the following:
+
+  *  **metadata.annotations.px/proxy-volume-nfs-subpath:** With the path to the sub-path directory. Volumes created from this PVC will only have access the sub-path, and none of the directories above it. <!-- can you create a new path here? or does it need to exist on the NFS share already? -->
```
Comment (adityadani) [inline content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md:141]: the annotation is
```
px/proxy-nfs-subpath: "foo/bar3"
```

Comment (adityadani) [reply content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md:141]: > 

good point.
 We should mention here:
- if the sub path does not exist, PX will create the sub path in the NFS share

File: content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md [hunk 9]
```diff
@@ -0,0 +1,213 @@
+---
+title: Create proxy volume PVCs
+weight: 4
+keywords: create proxy volumes, PVC, kubernetes, k8s
+description:
```
Comment (adityadani) [inline content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md:5]: We need to add a note saying this feature is available as a preview in 2.6.0-rc3 version.

File: content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md (hunk 8 above)
Comment (vinayakshnd) [reply content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md:141]: @adityadani we may also need to mention permissions required in order to successfully create subpath under NFS share, otherwise user may face errors like:

```
Events:
  Type     Reason              Age               From                         Message
  ----     ------              ----              ----                         -------
  Warning  ProvisioningFailed  6s (x2 over 21s)  persistentvolume-controller  Failed to provision volume with StorageClass "portworx-proxy-volume-volume": rpc error: code = Internal desc = Failed to create volume: unable to create a subpath foo at 70.0.46.165:/mnt: mkdir /var/lib/osd/proxy/create_tmp/pvc-3d23852b-c020-41dd-8cb5-1cede59aaf6d/foo: permission denied
  Warning  ProvisioningFailed  5s (x3 over 32s)  persistentvolume-controller  Failed to provision volume with StorageClass "portworx-proxy-volume-volume": rpc error: code = Internal desc = Failed to create volume: unable to create a subpath foo at 70.0.46.165:/mnt: mkdir /var/lib/osd/proxy/create_tmp/pvc-3d23852b-c020-41dd-8cb5-1cede59aaf6d/foo: permission denied
```

Comment (vinayakshnd) [reply content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md:141]: I was able to get around the above error by 
1. Giving `no_root_squash` in export options on the NFS server.
2. Running Nginx as a privileged container.

I am sure those would not be recommended steps. So can we document what user needs to give correct permissions for application to be able to perform writes?

File: content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+title: Create proxy volume PVCs
+weight: 4
+keywords: create proxy volumes, PVC, kubernetes, k8s
+description: Proxy an external data source onto a Portworx volume. 
+series: k8s-vol
+display: hidden
+---
+
+{{<info>}}
+**NOTE:** This feature is available as a preview in Portworx version 2.6.0-rc3.
```
Comment (adityadani) [inline content/portworx-install-with-kubernetes/storage-operations/create-pvcs/create-proxy-volume-pvcs.md:11]: we actually released 2.6.0-rc4
